pybikes/keolis.py

```diff
@@ -167,6 +167,55 @@ def update(self, scraper = None):
         # Update time as in 47 seconds ago: '47 secondes'
         self.extra['lastupd'] = xml_status.find('lastupd').text
 
+class IDEcycle(BikeShareSystem):
+
+    sync = True
+
+    def __init__(self, tag, meta, feed_url):
+        super(IDEcycle, self).__init__(tag, meta)
+        self.feed_url = feed_url
+
+    def update(self, scraper=None):
+        scraper = scraper or utils.PyBikesScraper()
+
+        stations = []
+        
+        html = scraper.request(self.feed_url)
+        STATIONS_RGX = "var\ geojsondatas\ =\ (.*?\}\]\})"
+        stations_html = re.findall(STATIONS_RGX, html, flags=re.DOTALL)        
+        data = json.loads(stations_html[0])
+        stations_data = data["features"]
+
+        for station_data in stations_data:
+            stations.append(IDEcycleStation(station_data))
+        self.stations = stations
+
+class IDEcycleStation(BikeShareStation):
+    def __init__(self, fields):
+        name = fields['properties']['title']
+        latitude = fields['geometry']['coordinates'][1]
+        longitude = fields['geometry']['coordinates'][0]
+
+        freeRGX = u"</li><li>(.*?)\ Places\ libres"
+        bikesRGX = u"<ul><li>(.*?)\ Vélos\ disponibles"
+        uidRGX = u"data-poi=\\\"(.*?)\\\"\ "
+        addressRGX = u"<p>(.*?)</br>"
+        slotsRGX = u"Capacité\ \:\ (.*?)\ vélos"
+        text = fields['properties']['popupContent']
```

Comment:
Let's keep regular expressions as a last resort for parsing html, or just for extracting contextual data but not markup. A common technique when text is not well structured is to just join all texts and extract it from there.

```
import re
from lxml import html

blergh = 'b>ALLEES </b><p>Avenue des Lilas</br>Capacité : 20 vélos</p><div class="informations"><ul><li>7 Vélos disponibles</li><li>13 Places libres</li></ul></div><a href="#" class="btn btn-link btn-sm text-primary btn_openpoi event_openpoi" data-poi="33" title="ALLEES ">+<span>d’infos</span></a>'

popup = html.fromstring(blergh)
text = " ".join(popup.xpath("//text()"))

free = re.findall(r'(\d+) Places libres', text, re.UNICODE)[0]
bikes = re.findall(r'(\d+) Vélos disponibles', text, re.UNICODE)[0]
slots = re.findall(r'Capacité : (\d+) vélos', text, re.UNICODE)[0]

uid = popup.xpath("//a/@data-poi")[0]

print(free, bikes, slots, uid)
```

```
> (u'13', u'7', u'20', '33')
```

It's also possible to just trust order of appearance instead of joining too, although that might backfire easily

```
# -*- coding: utf-8 -*-

import re
from lxml import html

blergh = 'b>ALLEES </b><p>Avenue des Lilas</br>Capacité : 20 vélos</p><div class="informations"><ul><li>7 Vélos disponibles</li><li>13 Places libres</li></ul></div><a href="#" class="btn btn-link btn-sm text-primary btn_openpoi event_openpoi" data-poi="33" title="ALLEES ">+<span>d’infos</span></a>'

popup = html.fromstring(blergh)

bikes, free = re.findall(r'(\d+)', ' '.join(popup.xpath("//ul/li/text()")))
slots, = re.findall(r'(\d+) vélos', ' '.join(popup.xpath("//p/text()")))
uid, = popup.xpath("//a/@data-poi")

print(free, bikes, slots, uid)
```

```
> (u'13', u'7', u'20', '33')
```


Note how on both cases `data-poi` can  be extracted cleanly without a regular expression

Reply:
Thanks a lot for your detailed comment, I learned a lot!!